packages/xforms-engine/src/parse/xpath/dependency-analysis.ts

```diff
@@ -0,0 +1,99 @@
+import { expressionParser } from '@getodk/xpath/expressionParser.js';
+import {
+	resolvePath,
+	resolvePredicateReference,
+	serializeNodesetReference,
+} from './path-resolution.ts';
+import { findPredicateReferences } from './predicate-analysis.ts';
+import { findLocationPathSubExpressionNodes, getPathExpressionNode } from './semantic-analysis.ts';
+
+export interface PathResolutionOptions {
+	/**
+	 * Excludes direct references to the nodeset specified as a context. This flag
+	 * has earlier precedent as `ignoreContextReference` in similar internal APIs.
+	 * The intent, in part, is to allow for certain in-spec exceptions to cycle
+	 * analysis (which we previously did explicitly, and foresee reintroducing in
+	 * some form relatively soon). As such, the default for this flag should be
+	 * overridden when a computation is (A) not among the core `<bind>`
+	 * computations or (B) is a computation which is explicitly expected to have
+	 * self-references (such as `constraint`).
+	 *
+	 * @default false
+	 *
+	 * @todo It is highly likely we will eliminate this in the future. We
+	 * considered removing its precursor (and existing code now referencing this)
+	 * in {@link https://github.com/getodk/web-forms/pull/67}, but ran into issues
+	 * with some of the reactive logic introduced there. It's likely that
+	 * subsequent improvements have addressed those issues, but it's also likely
+	 * that we will want similar special casing when we reintroduce explicit graph
+	 * cycle analysis.
+	 */
+	readonly ignoreReferenceToContextPath?: boolean;
+}
+
+const defaultPathResolutionOptions: Required<PathResolutionOptions> = {
+	ignoreReferenceToContextPath: false,
+};
+
+/**
+ * Resolves XPath paths referenced by an arbitrary expression to XForms
+ * `nodeset` references:
+ *
+ * 1. The provided {@link expression} is analyzed to find path sub-expressions
+ *    (LocationPaths, sub-expressions with a node-set producing FilterExpr).
+ *
+ * 2. When a {@link contextNodeset} is available, each sub-expression is
+ *    resolved to that context. See {@link resolvePath} for additional
+ *    detail.
+ *
+ * 3. Once resolved, each dependency path is further analyzed to identify
+ *    sub-expressions in any of its predicates, resolving each to their
+ *    respective predicates' Steps. See {@link resolvePredicateReference} for
+ *    additional detail.
+ */
+export const resolveDependencyNodesets = (
+	contextNodeset: string | null,
+	expression: string,
+	options: PathResolutionOptions = {}
+): readonly string[] => {
+	const { ignoreReferenceToContextPath } = {
+		...defaultPathResolutionOptions,
+		...options,
+	};
+
+	const contextNode = getPathExpressionNode(contextNodeset ?? '"not a path, null fallback"');
```

Comment:
Yeah, this was a little wonky and I think I meant to change it but forgot 😅. We do need to accept a null `contextNodeset` (and call it where that occurs[^1]), because the results can still be contextualized as the ancestor tree is resolved. In any case, as you've seen in the test suite, there is meaningful (albeit more limited) resolution logic applied even in the absence of a `contextNodeset`.

I think I'd originally done it this way to minimize the amount of branchy logic, but that was in an earlier pass which had... quite a bit more branchy logic in quite a few more places. (A huge chunk of the time spent on this change was focused on trying to tame the implementation complexity, so it could potentially be understood and maintained by others as well as my future self.) I think making it more explicitly conditional instead of this weird fallback is probably the most appropriate solution for now.

I do think it's likely we'll be able to make `contextNodeset` non-optional eventually, probably in the (er) context of fixing other aspects of groups without a `ref`.

[^1]: At a glance, I think the only current case is the expression parts of a `<label>` in a `<group>` without a `ref` attribute. And we have a variety of work to do to improve that case.